linkerd/proxy/dns-resolve/src/lib.rs

```diff
@@ -0,0 +1,156 @@
+        }
+    }
+
+    (adds, removes)
+}
+
+pub fn resolution_stream<S>(input: S) -> impl Stream<Item = Result<Update<SocketAddr>, Error>>
+where
+    S: futures::stream::Stream<Item = Result<Vec<SocketAddr>, dns::Error>>,
+    S: Send + 'static,
+{
+    try_stream! {
+        pin_mut!(input);
+        let mut current: Vec<SocketAddr> = Vec::new();
+        while let Some(result) = input.next().await {
+            let result = result?;
+            if result.is_empty() {
+                current.clear();
+                yield Update::Reset(Vec::new());
+            }
+
+            let (adds, removes) = diff_targets(&current, &result);
+            debug!(?adds, ?removes, "resolved");
+            if !adds.is_empty() {
+                let adds = adds.into_iter().map(|addr| (addr, addr)).collect();
+                yield Update::Add(adds);
+            }
+
+            if !removes.is_empty() {
+                yield Update::Remove(removes);
+            }
+            current.clear();
+            current.extend(result);
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use tokio::sync::mpsc;
+    use tokio_test::{assert_pending, assert_ready_eq, task};
+
+    #[tokio::test]
+    async fn yields_add() {
+        let (mut tx, rx1) = mpsc::channel(10);
+
+        let mut stream = task::spawn(resolution_stream(rx1).map_err(|_| ()));
+
+        let sa = "127.0.0.1:8080".parse().unwrap();
+        assert_pending!(stream.poll_next());
+        tx.send(Ok(vec![sa])).await.unwrap();
+        assert_ready_eq!(stream.poll_next(), Some(Ok(Update::Add(vec![(sa, sa)]))));
+    }
+
+    #[tokio::test]
+    async fn yields_remove() {
+        let (mut tx, rx1) = mpsc::channel(10);
+
+        let mut stream = task::spawn(resolution_stream(rx1).map_err(|_| ()));
+
+        let sa_1 = "127.0.0.1:8080".parse().unwrap();
+        let sa_2 = "127.0.0.2:8080".parse().unwrap();
+
+        tx.send(Ok(vec![sa_1])).await.unwrap();
+        assert_ready_eq!(
+            stream.poll_next(),
+            Some(Ok(Update::Add(vec![(sa_1, sa_1)])))
+        );
+        tx.send(Ok(vec![sa_2])).await.unwrap();
+        assert_ready_eq!(
+            stream.poll_next(),
+            Some(Ok(Update::Add(vec![(sa_2, sa_2)])))
+        );
+        assert_ready_eq!(stream.poll_next(), Some(Ok(Update::Remove(vec![sa_1]))));
+    }
+
+    #[tokio::test]
+    async fn yields_empty() {
+        let (mut tx, rx1) = mpsc::channel(10);
+
+        let mut stream = task::spawn(resolution_stream(rx1).map_err(|_| ()));
+
+        let sa_1 = "127.0.0.1:8080".parse().unwrap();
+        let sa_2 = "127.0.0.2:8080".parse().unwrap();
+
+        tx.send(Ok(vec![sa_1, sa_2])).await.unwrap();
+        assert_ready_eq!(
+            stream.poll_next(),
+            Some(Ok(Update::Add(vec![(sa_1, sa_1), (sa_2, sa_2)])))
+        );
+        tx.send(Ok(vec![])).await.unwrap();
+        assert_ready_eq!(stream.poll_next(), Some(Ok(Update::Empty)));
```

Comment:
Looks like this [doesn't compile](https://github.com/linkerd/linkerd2-proxy/pull/594/checks?check_run_id=1022713111#step:4:486) --- was it supposed to be `Update::Reset`?